### ISSUE TITLE
Multiple registration of workflows in a Webjobs environment

### ISSUE THREAD

**Issue by Helluva3ngineer** [danielgerlag/workflow-core#196]
Hello and thanks for your work on this awesome project.  I'm doing my first implementation with the framework and it's hosted as a set of Functions in a Webjob.  My initial workflow detects new user invitations and performs follow up steps once certain criteria are met.  I noticed today in my Application Insights logs the following:

_Workflow "xyz" version 1 is already registered_

I'm guessing this is a concurrency issue where two threads might be doing the same thing at the same time.  The Function is Timer based and I have it decorated as follows:

[Singleton(Mode = SingletonMode.Listener)]

...just curious if anyone has attempted something like this before and if there are any best practices?  I'm thinking about trying to acquire my own Mutex so that I have better control but really feel this should be working as is.  Thanks!

**Comment by danielgerlag** [comment]
Just to clarify, are you using Azure Functions or WebJobs within Azure App Service?

**Comment by Helluva3ngineer** [comment]
Correct.  It's hosted as a webjob under an ASP.NET Core site. The Webjob itself is a .NET Core 2.1 console application.

**Comment by danielgerlag** [comment]
You only need to register your workflows once, when the console app starts. Best place to do it is right after configuring your IoC container.

**Comment by cristiamz** [comment]
Hi Daniel, I'm doing something similar, but I'm using json workflows that will get executed each time a message on a Azure Service bus is received. The issue I'm having is that when the requests are close to each other, the steps get executed multiple times per execution. 

Here is the code I have right now 


        public async Task ExecuteWorkflowAsync(dynamic DataObject, string WorkflowDef) {

            IServiceProvider serviceProvider = ConfigureServices();

            //start the workflow host
            var Workflowhost = serviceProvider.GetService<IWorkflowHost>();
            var _registry = serviceProvider.GetService<IWorkflowRegistry>();
            var loader = serviceProvider.GetService<IDefinitionLoader>();

            loader.LoadDefinition(WorkflowDef);
            Workflowhost.Start();

            dynamic wfDef = JsonConvert.DeserializeObject(WorkflowDef);
            string Id = wfDef.Id;
            int Version = wfDef.Version;

            var def = _registry.GetDefinition(Id, Version);

            if ((DataObject != null) && (def.DataType != null))
            {
                var dataStr = JsonConvert.SerializeObject(DataObject);
                object dataObj = JsonConvert.DeserializeObject(dataStr, def.DataType);
                var Workflowid = await Workflowhost.StartWorkflow(workflowId: Id, version: Version, data: dataObj);
            }

            await Task.Delay(5000);
            Workflowhost.Stop();

        }

        private static IServiceProvider ConfigureServices()
        {
            //setup dependency injection
            IServiceCollection services = new ServiceCollection();
            services.AddLogging();
            services.AddWorkflow();
            
            var serviceProvider = services.BuildServiceProvider(true);
            serviceProvider.CreateScope();

            return serviceProvider;
        }



My intention is to use your library as a one time execution, I send the json workflow def on the parameter WorkflowDef and the object assigned to that workflow on DataObject. 

Also notice that I had to put that  await Task.Delay(5000); to wait some time for the workflow to execute and then stop the host because it kept running and the next time I run the workflow it duplicate some steps. 

Let me know if you have any Ideas

Thanks!
